src/neural/blas/batchnorm.cc

```diff
@@ -0,0 +1,64 @@
+/*
+ This file is part of Leela Chess Zero.
+ Copyright (C) 2018 The LCZero Authors
+
+ Leela Chess is free software: you can redistribute it and/or modify
+ it under the terms of the GNU General Public License as published by
+ the Free Software Foundation, either version 3 of the License, or
+ (at your option) any later version.
+
+ Leela Chess is distributed in the hope that it will be useful,
+ but WITHOUT ANY WARRANTY; without even the implied warranty of
+ MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ GNU General Public License for more details.
+
+ You should have received a copy of the GNU General Public License
+ along with Leela Chess.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "batchnorm.h"
+
+#include <cmath>
+
+namespace lczero {
+
+void Batchnorm::Apply(const size_t batch_size, const size_t channels,
+                      float* data, const float* means, const float* stddivs,
+                      const float* eltwise) {
+  for (size_t i = 0; i < batch_size; i++) {
+    for (size_t c = 0; c < channels; ++c) {
+      auto mean = means[c];
+      auto scale_stddiv = stddivs[c];
+
+      if (eltwise == nullptr) {
+        // Classical BN
+        auto arr = &data[c * kSquares];
+        for (size_t b = 0; b < kSquares; b++) {
+          float val = scale_stddiv * (arr[b] - mean);
+          arr[b] = val > 0 ? val : 0;
+        }
+      } else {
+        // BN + residual add
+        auto arr = &data[c * kSquares];
+        auto res = &eltwise[c * kSquares];
+        for (size_t b = 0; b < kSquares; b++) {
+          float val = res[b] + (scale_stddiv * (arr[b] - mean));
+          arr[b] = val > 0 ? val : 0;
+        }
+      }
+    }
+    data += channels * kSquares;
+    if (eltwise != nullptr) eltwise += channels * kSquares;
+  }
+}
+
+void Batchnorm::OffsetMeans(std::vector<float>& bn_means,
```

Comment:
Output parameters should be pointers, not references. And be the last in the list of arguments.

Reply:
I still don't understand that style guidance. The whole point of references was to be able to make inplace modifications of this nature without the risk of munging the address caller-side. (The easier syntax doesn't hurt either.)

Reply:
OffsetMeans and InvertStddev are special cases that I did not touch.

No we cannot use vectors as the length parameter would be meaningless.

Reply:
@dubslow 
Many style guides (including google styleguide that we happen to use) disallow mutable references because otherwise it can lead to surprises at calling point (that's a habit from C language, where there's no references). E.g. (extreme example):
```cpp
int x = 32;
ComputeSomethingXTimes(x);  // Unexpected that x may change.
DoSomethingElseXTimes(x);  // x is 0 now.

void ComputeSomethingXTimes(int& x) {
   while (x--) ComputeSomething();
}
```

Reply:
Still not fixed.